t/app/controller/waste.t

```diff
@@ -257,14 +380,709 @@ FixMyStreet::override_config {
+        $mech->get_ok('/waste/12345');
+        $mech->content_lacks('Garden Waste', "no mention of Garden Waste");
+        $mech->content_lacks('/waste/12345/garden_renew', "no Garden Waste renewal link");
+
+        $mech->log_in_ok($user->email);
+        $mech->get_ok('/waste/12345/garden_renew');
+        $mech->submit_form_ok({ with_fields => {
+            current_bins => 1,
+            new_bins => 0,
+            payment_method => 'credit_card',
+        } });
+        $mech->content_contains('20.00');
+        $mech->submit_form_ok({ with_fields => { tandc => 1 } });
+        is $sent_params->{amount}, 2000, 'correct amount used';
+
+        my $url = $sent_params->{returnUrl};
+        my ($report_id) = ( $url =~ m#/(\d+)$# );
+        my $new_report = FixMyStreet::DB->resultset('Problem')->find( { id => $report_id } );
+
+        is $new_report->category, 'New Garden Subscription', 'correct category on report';
+        is $new_report->get_extra_field_value('payment_method'), 'credit_card', 'correct payment method on report';
+        is $new_report->get_extra_field_value('Subscription_Details_Quantity'), 1, 'correct bin count';
+        is $new_report->get_extra_field_value('Container_Request_Details_Action'), '', 'no container request action';
+        is $new_report->get_extra_field_value('Container_Request_Details_Quantity'), '', 'no container request count';
+        is $new_report->state, 'unconfirmed', 'report not confirmed';
+
+        $new_report->discard_changes;
+        is $new_report->get_extra_metadata('scpReference'), '12345', 'correct scp reference on report';
+
+        $mech->get_ok('/waste/pay_complete/' . $new_report->id);
+        is $sent_params->{scpReference}, 12345, 'correct scpReference sent';
+
+        $new_report->discard_changes;
+        is $new_report->state, 'confirmed', 'report confirmed';
+        is $new_report->get_extra_metadata('payment_reference'), '54321', 'correct payment reference on report';
+    };
+
+    subtest 'cancel credit card sub' => sub {
+        set_fixed_time('2021-03-09T17:00:00Z'); # After sample data collection
+        $mech->log_in_ok($user->email);
+        $mech->get_ok('/waste/12345/garden_cancel');
+        $mech->submit_form_ok({ with_fields => { confirm => 1 } });
+
+        my $new_report = FixMyStreet::DB->resultset('Problem')->search(
+            { user_id => $user->id },
+            { order_by => { -desc => 'id' } },
+        )->first;
+
+        is $new_report->category, 'Cancel Garden Subscription', 'correct category on report';
+        is $new_report->get_extra_field_value('Container_Request_Details_Action'), 2, 'correct container request action';
+        is $new_report->get_extra_field_value('Container_Request_Details_Quantity'), 1, 'correct container request count';
+        is $new_report->state, 'confirmed', 'report confirmed';
+    };
+
+    subtest 'check new sub credit card payment lookup failed' => sub {
```

Comment:
This and next test seem identical apart from the mock - refactor? Also, they're testing what the code does but I'm not sure that's a great outcome for the user at present